Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/test-settings/test-settings.component.ts

```diff
@@ -132,18 +138,36 @@ export class TestSettingsComponent implements OnInit {
      * @param testObject is an object of class Test
      */
     launchTestDialog(id: number, testObject: Test) {
-        this.testService.updateTestById(id, testObject).subscribe((response) => {
-            this.openSnackBar(this.testSettingsUpdatedMessage);
-            let instance = this.dialog.open(TestLaunchDialogComponent).componentInstance;
-            instance.testSettingObject = testObject;
-        },
-            errorHandling => {
-                this.response = errorHandling.json();
-                this.errorMessage = this.response['error'];
-                this.snackbarRef.open(this.errorMessage, 'Dismiss', {
-                    duration: 3000,
-                });
-            },
-        );
+        let categories = this.testDetails.categoryAcList.some(function (x) {
+            return x.isSelect;
+        });
+        if (categories) {
+            let questions = this.testDetails.categoryAcList.some(function (x) {
+                return (x.numberOfSelectedQuestion !== 0);
+            });
+            if (questions) {
+                this.testService.updateTestById(id, testObject).subscribe((response) => {
+                    this.openSnackBar(this.testSettingsUpdatedMessage);
+                    let instance = this.dialog.open(TestLaunchDialogComponent).componentInstance;
+                    instance.testSettingObject = testObject;
+                },
+                    errorHandling => {
+                        this.response = errorHandling.json();
+                        this.errorMessage = this.response['error'];
+                        this.snackbarRef.open(this.errorMessage, 'Dismiss', {
+                            duration: 3000,
+                        });
+                    },
+                );
+            }
+            else {
+                this.testDetails.isQuestionMissisng = true;
+                let dialogRef = this.dialog.open(IncompleteTestCreationDialogComponent, { data: this.testDetails });
+            }
+        }
+        else {
+            this.testDetails.isQuestionMissisng = false;
+            let dialogRef = this.dialog.open(IncompleteTestCreationDialogComponent, { data: this.testDetails });
```

Comment:
why two times component is called in diff if conditions. inputs are similar for both the components. you should optimise code.